### DISCUSSION TITLE
Is overriding/extending the default scalar Coercing implementations supported?

### DISCUSSION THREAD

**Discussion by oryan-block** [graphql-java/graphql-java#3173] [Q&A] [answered]
This seems to accidentally work for `ID` and `Int` but not for `Boolean` (or `String`) because `IncludeDirective` and `SkipDirective` have an `if` argument that's using `GraphQLBoolean` for its type (same for the introspection types). So then when I'm building a schema with my own implementation of the boolean coercing I'm getting the error:

> All types within a GraphQL schema must have unique names. No two provided types may have the same name.
No provided type may have a name which conflicts with any built in types (including Scalar and Introspection types).
You have redefined the type 'Boolean' from being a 'GraphQLScalarType' to a 'GraphQLScalarType'

Is this a bug or is it just not supported? 

```
    GraphQLScalarType.newScalar()
        .name("Boolean") <-- can't create this because it's already created by the library
        .coercing(...)
        .build();
```

**Comment by bbakerman** [comment, marked as answer]
Providing your own standard scalars is not intended to be supported.  I am suprised it works even for ID and Int to be honest

Code can statically reference `graphql.Scalars#GraphQLInt` and so on at any time and  its meant to be a canonical representation of the inbuilt graphql schemas

So in short we dont intend for people to replace them 

That said - what is your use case for replacing the standard scalars with new implementations?

**Reply by oryan-block** [reply]
Well we use an ID coercing implementation to sanitize possible garbage input.
For the `Boolean` and `Int` case I wanted to provide a temporary ability to parse strings (as was previously possible) until we get all clients up to date.
Maybe a future solution could involve some kind of decorator?

**Reply by bbakerman** [reply]
I had a quick look at this to see if I could transform the schema to replace the type

```
def "can extend built in scalars"() {

        def sdl = """
            type Query {
                booleanField : Boolean
            }
        """

        def schema = TestUtil.schema(sdl)

        GraphQLScalarType customBoolean = GraphQLScalarType.newScalar()
                .name("Boolean")
                .coercing(new Coercing<Object, Object>() {
                    @Override
                    Object serialize(@NotNull Object dataFetcherResult, @NotNull GraphQLContext graphQLContext, @NotNull Locale locale) throws CoercingSerializeException {
                        return "customTrue"
                    }

                    @Override
                    Object parseValue(@NotNull Object input, @NotNull GraphQLContext graphQLContext, @NotNull Locale locale) throws CoercingParseValueException {
                        if (input == "customTrue") {
                            return true
                        }
                        return Scalars.GraphQLBoolean.coercing.parseValue(input, graphQLContext, locale)
                    }

                    @Override
                    Object parseLiteral(@NotNull Value<?> input, @NotNull CoercedVariables variables, @NotNull GraphQLContext graphQLContext, @NotNull Locale locale) throws CoercingParseLiteralException {
                        return Scalars.GraphQLBoolean.coercing.parseLiteral(input, variables, graphQLContext, locale)
                    }

                    @Override
                    Value<?> valueToLiteral(@NotNull Object input, @NotNull GraphQLContext graphQLContext, @NotNull Locale locale) {
                        return Scalars.GraphQLBoolean.coercing.valueToLiteral(input, graphQLContext, locale)
                    }
                })
                .build()

        def transformedSchema = new SchemaTransformer().transform(schema, new GraphQLTypeVisitorStub() {
            @Override
            TraversalControl visitGraphQLScalarType(GraphQLScalarType node, TraverserContext<GraphQLSchemaElement> context) {
                if (node.name.equals("Boolean")) {
                    return TreeTransformerUtil.changeNode(context,customBoolean)
                }
                return TraversalControl.CONTINUE
            }
        })


        GraphQL graphQL = GraphQL.newGraphQL(transformedSchema).build()

        def query = ExecutionInput.newExecutionInput("""
                query q {
                    booleanField  
                """)
                .root(["booleanField": true])
                .build()

        when:
        def er = graphQL.execute(query)
        then:
        er.data == ["booleanField": "customTrue"]
    }
```

In theory this should  BUT I think the edit here is NOT replacing all scalars like it should and elements such as magic directives are not being replaced (which might be done for good reasons)

I was just trying to see if this would succeed or not.  And it doesnt - I posted this just to show my investigation

**Reply by bbakerman** [reply]
One of the challenges here with replacing the built in scalars (alongside  the client usage semantics) is that object equality is important to the schema  - there is meant to be one implementation of any given type within a schema - including scalars

However we have code that addresses well known scalars by well know static addresses - like you  have seen

```
public static final GraphQLDirective IncludeDirective = GraphQLDirective.newDirective()
            .name("include")
            .description("Directs the executor to include this field or fragment only when the `if` argument is true")
            .argument(newArgument()
                    .name("if")
                    .type(nonNull(GraphQLBoolean))
                    .description("Included when true."))
            .validLocations(FRAGMENT_SPREAD, INLINE_FRAGMENT, FIELD)
            .build();
```

So code can not only access static scalars via `graphql.Scalars#GraphQLBoolean` but they can access static objects that are themselves referencing the built in ones

So the code is fragile if we say that built in scalars can change.  All the code that assumes it has access to a static JVM wide single implementation starts to break

**Reply by oryan-block** [reply]
Thanks for looking into this. I suppose enabling this feature will require more breaking changes then. So might not be worth it? But clearing up this issue in the documentation would be a good idea I think.
